### DISCUSSION TITLE
Issue uninstalling Toltec

### DISCUSSION THREAD

**Discussion by murchu27** [toltec-dev/toltec#893] [Support] [answered]
Hey gang! Thanks for your continued work on this great project!

I used to use Toltec and 3rd party apps a lot, but at some point I stripped them away to let a friend use the tablet.
Now, I see that I have let my tablet update a few too many times, and my version (3.8.2) is currently not supported by Toltec.
No harm there, I know that you folks will get Toltec updated before too long!

In the meantime though, I wanted to check my tablet to make sure that I cleaned out Toltec properly, as I used to do some right hack jobs on these kinds of projects. Sure enough, it seems there's still traces of Toltec on my system.

When I try to run `toltecctl uninstall`, I get `toltecctl: command not found`. Seems good so far.
But then, when I check directories in `/home/root/`, I see an `.entware` directory. Uh oh!
I also still have the `opt.mount` service running on the tablet, visible with `systemctl status opt.mount --no-pager`.
If I try to stop this service, it fails with the following logs:

```
Jun 07 12:39:17 reMarkable systemd[1]: Unmounting Bind mount /home/root/.entware over /opt...
Jun 07 12:39:17 reMarkable umount[553]: umount: /opt: target is busy.
Jun 07 12:39:17 reMarkable systemd[1]: opt.mount: Mount process exited, code=exited, status=32/n/a
Jun 07 12:39:17 reMarkable systemd[1]: Failed unmounting Bind mount /home/root/.entware over /opt.
```

On a normal Linux box, I might run `lsof` to see what's using `/opt`, but the reMarkable shell doesn't seem to have `lsof`...

Can someone help me get Toltec fully removed from my tablet, so that it is nice and clean for when Toltec supports it?

**Comment by jkesselm** [comment]
There is a tool -- I'm away from my machine and don't remember the name offhand -- that will let you switch to any of a fairly wide range of versions of the reMarkable base software. I had to use that to roll back to a version Toltec supported. So I'd suggest trying that, then doing a Toltec reinstall.

**Reply by murchu27** [reply]
Ah yes, I think that's ddvk/remarkable-update: https://github.com/ddvk/remarkable-update

I could try that... My only hesitation is that, even if I'm on a supported reMarkable version and could install Toltec, I will probably get an error about Toltec being installed already, because of the existence of the `.entware` directory and `opt.mount` service. So either way, I think I will still need to manually uninstall Toltec somehow.

**Reply by Eeems** [reply]
https://remarkable.guide/faqs.html#can-i-downgrade-to-a-different-os-version
has a link to the latest recommended tool. After doing a downgrade/upgrade, you just need to delete `/home/root/.entware` to remove toltec, as no changes to the root partition are there anymore (https://remarkable.guide/faqs.html#what-happens-to-third-party-software-when-i-update-or-downgrade-my-os) which is the reason you need to reenable after an upgrade.

**Comment by Eeems** [comment, marked as answer]
Are you sure you have toltec installed and not [remarkable_entware](https://github.com/Evidlo/remarkable_entware)?

If they did somehow install toltec on 3.8, which is not supported, you will encounter issues if you tried to update to the latest version that will probably leave you in a soft-bricked state.

You can install lsof with `opkg install lsof`, and then see what is running that is using `/opt`. At which time you'll want to stop/disable the services. That said, I would recommend just downgrading/upgrading twice, and then removing the `/home/root/.entware` folder.
https://remarkable.guide/faqs.html#can-i-downgrade-to-a-different-os-version
https://remarkable.guide/faqs.html#what-happens-to-third-party-software-when-i-update-or-downgrade-my-os

**Reply by murchu27** [reply]
What's the purpose of downgrading/upgrading, if I am just removing `/home/root/.entware`? Can't I just do that without downgrading/upgrading? Just trying to fully understand the steps before I do anything destructive.

**Reply by Eeems** [reply]
Downgrading/upgrading replaces the root partition with a known state, as you don't know what modifications have been made to it. This way you avoid soft-bricking that would be caused by changes to the root partition that no longer work if the underlying files are now missing.

This is the same reason you don't do a "factory reset" with third party software installed, and the instructions for doing a true factory reset involve downgrading/upgrading twice before using the built-in factory reset. While they do instruct you to uninstall third party software first (like toltec), you can also skip that step and just do two system upgrades/downgrades in a pinch.
https://remarkable.guide/faqs.html#is-it-safe-to-factory-reset-my-device
https://remarkable.guide/tech/factory-reset.html

**Reply by murchu27** [reply]
Okay got it, thanks for the explanation! I'll try out `codexctl`, and post the results here.

**Reply by murchu27** [reply]
Alright, played around with this and took a lot of notes.

Before starting, my tablet had 3.8 and 3.11 installed, with 3.8 being the active installation, and 3.11 on the fallback partition.
I downloaded the most recent pre-compiled binary of `codexctl` and SFTP'd to my tablet.
After marking `codexctl.bin` as executable (`chmod +x codexctl.bin`), I simply ran `./codexctl.bin install 3.3.2.1666` to install version 3.3.2.1666.
The tablet rebooted without issue into v3.3, and I immediately switched off auto-updates.

I was then keen to get rid of 3.11, as I've seen lots of 3rd party stuff (notably rmfakecloud) has trouble with it.
But as noted in a few places on the `codexctl` repo, the same `install` commands don't work when running on v3.11.
So, instead, I downloaded v3.8 to my computer, extracted the image, and transferred to my tablet.

```
./codexctl.bin download 3.8.2.1965
./codexctl.bin extract ~/Downloads/3.8.2.1965_reMarkable2-D1pFxylu6p.signed
mv extracted reMarkable2-3.8.2.1965.img
# SFTP to tablet...
```

Then, back on the tablet, after checking [which partition was my fallback partition](https://github.com/Jayy001/codexctl/issues/71#issuecomment-2099619757), I overwrote that partition with the extracted v3.8 image: `dd if=reMarkable2-3.8.2.1965.img of=/dev/mmcblk2p2`

After doing this, I switched to the newly written v3.8 partition using `./codexctl.bin restore`, to make sure it was working. When I did this, my tablet seemed to lock up (display was frozen, couldn't SSH). But as soon as I plugged in a USB cable, it rebooted automatically, and booted without issue into the v3.8 partition.

I once again turned off auto-updates as soon as I booted in.

From what I can tell, this should constitute "two system upgrades/downgrades", as both partitions were overwritten with a fresh installation. So then I removed `.entware`. And checking `systemctl status opt.mount` shows that `opt.mount` is now gone! So looks like a full uninstall!

@Eeems greatly appreciate your help with this :)
